packages/utils/src/ast-utils/helpers.ts

```diff
@@ -40,19 +40,18 @@ export const isNodeOfTypeWithConditions = <
 
 export const isTokenOfTypeWithConditions = <
   TokenType extends AST_TOKEN_TYPES,
-  ExtractedToken extends Extract<TSESTree.Token, { type: TokenType }>,
   Conditions extends Partial<TSESTree.Token & { type: TokenType }>,
 >(
   tokenType: TokenType,
   conditions: Conditions,
 ): ((
   token: TSESTree.Token | null | undefined,
-) => token is Conditions & ExtractedToken) => {
+) => token is Conditions & Extract<TSESTree.Token, { type: TokenType }>) => {
   const entries = Object.entries(conditions) as ObjectEntries<TSESTree.Token>;
 
   return (
     token: TSESTree.Token | null | undefined,
-  ): token is Conditions & ExtractedToken =>
```

Comment:
Some of these cases were intentional to make the code nicer - as  a way of "extracting" the type to a shared name for clarity and deduplication.

It shows that the "golden rule" isn't really an "always" thing.
In this case as well - the generic is used twice - shouldn't this be ignored by the rule?
I think that the new code is less readable.

This is *technically* unsafe if, say, someone manually passed the type parameter. But realistically:
- there are multiple complex type params that are intended to be inferred, not declared
- the type param's constraint is derived from another type param

I'm torn here... should we have an opt-out for this case? Or just rely on disable comments?

Reply:
> the generic is used twice

Is it? It's used in the return type of the returned function. It's in the AST twice only because the function has an explicit return type. If we used an inferred type, it'd be in the AST only once.


> opt-out

Hmm, what would the opt-out be on?

> disable comments

I'll go with that for now, I can see why this is useful.

Reply:
> > the generic is used twice
> 
> Is it? It's used only in the return type of the returned function. It's in the AST twice only because the function has an explicit return type. If we used an inferred type, it'd be in the AST only once.

This rule is about whether a type parameter is being used to relate two types. The implementation isn't relevant to that, unless there's an inferred return type. This rule is just about type signatures.

I think the specific issue we're running into here is interesting, but also pretty unusual:

- If you want to name a type expression that only appears in the implementation, you can define a type alias in the function body.
- If you're using the optional type parameter trick to factor out a type expression that appears 2+ times in the function signature, then it will not violate this rule.

So you only run into this situation if you're using the trick to factor out a type expression that appears exactly once in the function signature and one or more times in the body.

It's a shame that TS doesn't have a better way to define scoped type aliases without polluting the function signature (https://github.com/microsoft/TypeScript/issues/42388). But given that, I don't find this duplication very offensive, since it eliminates implementation details from the public type signature. On the other hand, I also don't find the `eslint-disable` very offensive since I think this is kind of a weird edge case.

I assume there's a typescript-eslint policy against letting `isTokenOfTypeWithConditions` have an inferred return type? That would eliminate the duplication and the need for `ExtractedToken` type parameter.

Reply:
There is, we keep the [@typescript-eslint/explicit-function-return-type](https://typescript-eslint.io/rules/explicit-function-return-type) rule enabled.

Reply:
We could probably relax that.
I'd be okay with us doing `isolatedDeclarations` instead of EFRT.

Reply:
I like it! #9419